block/manager.go

```diff
@@ -101,11 +101,23 @@ type Manager struct {
 
 // getInitialState tries to load lastState from Store, and if it's not available it reads GenesisDoc.
 func getInitialState(store store.Store, genesis *cmtypes.GenesisDoc) (types.State, error) {
+	b, err := store.GetBlock(uint64(genesis.InitialHeight))
+	if err != nil {
+		// we've never seen a block at InitialHeight.
+		// assume we're syncing from a trusted snapshot
+		return types.NewFromGenesisDoc(genesis)
+	}
+	if !bytes.Equal(b.SignedHeader.AppHash, genesis.AppHash.Bytes()) {
+		// the user-supplied genesis overrides a previously-seen block
+		// must be a hard-fork.
+		return types.NewFromGenesisDoc(genesis)
```

Comment:
imho, we shouldn't randomly start over-writing the state. instead, we should assume that node is restarted with hard-fork state and the corresponding InitialHeight is provided to resume the node.